src/Query.tsx

```diff
@@ -0,0 +1,191 @@
+
+  Object.keys(fields).forEach(key => {
+    if (typeof fields[key] === 'function') {
+      fields[key] = fields[key].bind(observable);
+    }
+  });
+
+  return fields;
+}
+
+class Query extends React.Component<QueryProps, QueryState> {
+  private client: ApolloClient<any>;
+  private queryObservable: ObservableQuery<any>;
+  private querySubscription: ZenObservable.Subscription;
+
+  static contextTypes = {
+    client: PropTypes.object.isRequired,
+  };
+
+  constructor(props, context) {
+    super(props, context);
+
+    invariant(
+      !!context.client,
+      `Could not find "client" in the context of Query. Wrap the root component in an <ApolloProvider>`,
+    );
+    this.client = context.client;
+
+    this.initializeQueryObservable(props);
+    this.state = {
+      result: this.queryObservable.currentResult(),
+    };
+  }
+
+  componentDidMount() {
+    this.startQuerySubscription();
+  }
+
+  componentWillReceiveProps(nextProps, nextContext) {
+    if (
+      shallowEqual(this.props, nextProps) &&
+      this.client === nextContext.client
+    ) {
+      return;
+    }
+
+    if (this.client !== nextContext.client) {
+      this.client = nextContext.client;
+    }
+    this.removeQuerySubscription();
+    this.initializeQueryObservable(nextProps);
+    this.startQuerySubscription();
+    this.updateCurrentData();
+  }
+
+  componentWillUnmount() {
+    this.removeQuerySubscription();
+  }
+
+  private initializeQueryObservable = props => {
+    const {
+      variables,
+      pollInterval,
+      fetchPolicy,
+      notifyOnNetworkStatusChange,
+      query,
+    } = props;
+
+    const operation = parser(query);
+
+    invariant(
+      operation.type === DocumentType.Query,
+      `The <Query /> component requires a graphql query, but got a ${operation.type ===
+      DocumentType.Mutation
+        ? 'mutation'
+        : 'subscription'}.`,
+    );
+
+    const clientOptions = {
+      variables,
+      pollInterval,
+      query,
+      fetchPolicy,
+      notifyOnNetworkStatusChange,
+    };
+
+    this.queryObservable = this.client.watchQuery(clientOptions);
+  };
+
+  private startQuerySubscription = () => {
+    this.querySubscription = this.queryObservable.subscribe({
+      next: this.updateCurrentData,
+      error: this.updateCurrentData,
+    });
+  };
+
+  private removeQuerySubscription = () => {
+    if (this.querySubscription) {
+      this.querySubscription.unsubscribe();
+    }
+  };
+  private updateCurrentData = () => {
+    this.setState({ result: this.queryObservable.currentResult() });
+  };
+
+  getRenderProps = () => {
+    const { result } = this.state;
+
+    const { loading, error, networkStatus, data } = result;
+
+    const renderProps = {
+      data,
+      loading,
+      error,
+      networkStatus,
+      ...observableQueryFields(this.queryObservable),
+    };
+
+    return renderProps;
+  };
+
+  render() {
+    const { children } = this.props;
+    const result = this.getRenderProps();
```

Comment:
I would just put `getRenderProps` implementation into render.  Alternatively you could keep it and rename it to `getResult` to be consistent.